lib/hal_client/link.rb

```diff
@@ -0,0 +1,63 @@
+require 'hal_client/representation'
+
+class HalClient
+
+  # HAL representation of a single link. Provides access to an embedded representation.
+  class Link
+
+    # Create a new Link
+    #
+    # options - name parameters
+    #   :rel - This Link's rel property
+    #   :target - An instance of Representation
+    #   :template = A URI template ( https://www.rfc-editor.org/rfc/rfc6570.txt )
+    def initialize(options)
+      @rel = options[:rel]
+      @target = options[:target]
+      @template = options[:template]
+
+      (fail ArgumentError, "A rel must be provided") if @rel.nil?
+
+      if @target.nil? && @template.nil?
+        (fail ArgumentError, "A target or template must be provided")
+      end
+
+      if @target && @template
+        (fail ArgumentError, "Cannot provide both a target and a template")
+      end
+
+      if @target && !@target.kind_of?(Representation)
+        fail InvalidRepresentationError, "Invalid HAL representation: #{target.inspect}"
+      end
+
+      if @template && !@template.kind_of?(Addressable::Template)
+        fail InvalidRepresentationError, "Invalid Addressable::Template: #{template.inspect}"
+      end
+    end
+
+    attr_accessor :rel, :target, :template
+
+    def href
+      templated? ? template.pattern : target.href
+    end
```

Comment:
i think this method should return a `Addressable::Template` object when the link is templated. that would make it behave like `#related_hrefs`. Or we could rename this method to `raw_href`.